test/NuGet.Core.Tests/NuGet.Configuration.Test/SearchTreeTests.cs

```diff
@@ -0,0 +1,344 @@
+            var packageSourcesMatchFull1 = configuration.GetConfiguredPackageSources("Contoso.Opensource.MVC");
+            Assert.Equal(1, packageSourcesMatchFull1.Count);
+            Assert.Equal("publicrepository", packageSourcesMatchFull1.First());
+
+            // Match
+            var packageSourcesMatchFull2 = configuration.GetConfiguredPackageSources("Contoso.Opensource.MVC.ASP");
+
+            Assert.Equal(1, packageSourcesMatchFull2.Count);
+            Assert.Equal("publicrepository", packageSourcesMatchFull2.First());
+
+            // No match
+            var packageSourcesNoMatch = configuration.GetConfiguredPackageSources("random");
+            Assert.Null(packageSourcesNoMatch);
+        }
+
+        [Fact]
+        public void SearchTree_WithMultipleSources()
+        {
+            // Arrange
+            using var mockBaseDirectory = TestDirectory.Create();
+            var configPath1 = Path.Combine(mockBaseDirectory, "NuGet.Config");
+            SettingsTestUtils.CreateConfigurationFile(configPath1, @"<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+    <packageNamespaces>
+        <packageSource key=""nuget.org"">
+            <namespace id=""stuff"" />
+        </packageSource>
+        <packageSource key=""contoso"">
+            <namespace id=""moreStuff"" />
+        </packageSource>
+        <packageSource key=""privateRepository"">
+            <namespace id=""private*"" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>");
+            var settings = Settings.LoadSettingsGivenConfigPaths(new string[] { configPath1 });
+
+            // Act & Assert
+            var configuration = PackageNamespacesConfiguration.GetPackageNamespacesConfiguration(settings);
+            Assert.True(configuration.IsNamespacesEnabled);
+            configuration.Namespaces.Should().HaveCount(3);
+
+            var packageSourcesMatchFull1 = configuration.GetConfiguredPackageSources("stuff");
+
+            Assert.Equal(1, packageSourcesMatchFull1.Count);
+            Assert.Equal("nuget.org", packageSourcesMatchFull1.First());
+
+            var packageSourcesMatchPartial1 = configuration.GetConfiguredPackageSources("stu");
+            Assert.Null(packageSourcesMatchPartial1);
+
+            var packageSourcesMatchFull2 = configuration.GetConfiguredPackageSources("moreStuff");
+            Assert.Equal(1, packageSourcesMatchFull2.Count);
+            Assert.Equal("contoso", packageSourcesMatchFull2.First());
+
+            var packageSourcesMatchPartial2 = configuration.GetConfiguredPackageSources("PrivateTest");
+            Assert.Equal(1, packageSourcesMatchPartial2.Count);
+            Assert.Equal("privaterepository", packageSourcesMatchPartial2.First());
+
+            var packageSourcesNoMatch = configuration.GetConfiguredPackageSources("random");
+            Assert.Null(packageSourcesNoMatch);
+        }
+
+        [Fact]
+        public void SearchTree_WithMultipleSourcesMultiparts()
```

Comment:
I am now seeing that you refer to ids or namespaces with `.` as multipart. 

I'm not sure that's relevant here. `.` is not special cased anywhere in the code.
Or at least shouldn't be.

Reply:
I just tried to cover all possible input cases. Maybe it's not important now, but in the future if we start implementing `strict` mode or other futures then we can be confident nothing is broken.